### PR TITLE
3D Unet for Segmentation Of HipMRI data

### PR DESCRIPTION
**Pull request by WilliamMercado**
The goal of this is to segment MRI data, specifically regarding the area around the prostate. I used a standard design of a 3D UNet as described in “U-Net: Convolutional Networks for Biomedical Image Segmentation.” 
General Details about usage and setup for testing included in README.md in the 3D-Unet folder.

Included Files:
1.  Informational
    - README.md : General information about model usage and setup
    - ExampleResults.png : Example Generation Results (Quite disappointing)
    - Test_Train_loss.png : Example Test/Train loss results (Again, quite disappointing)
    - .gitignore : prevent committing of un-necessary files within this folder
2. Execution
    - const.py : Various important constants concerning the location of saved information including dataset location (Details in file)
    - dataset.py : Classes and functions for loading, splitting and creating batches for the dataset.
    - modules.py : Classes for creating the UNet model
    - util.py : Utility functions for augmentation of the dataset. Mainly Rotating 3D images.
    - train.py : runtime program for training a model.
    - predict.py : example program for testing a trained model and showing its output

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**

Recognition problem: total 7 Marks

Problem solving: (1)
•	Segmentation results outputted with poor performance.
•	Dice score and individual class results not provided for model evaluation.
•	Segmentation output was not properly processed for output display (for the segmentation output, each pixel should be classified as one of the 6 classes).
Algorithm implementation: (3)
•	Algorithm implemented in PyTorch for the UNet3D model and functions.
Design: Modularised and reusable code (1)
Commenting: Sufficient comments throughout the code (1)
Difficulty: (1)
•	3D UNet – Normal 
•	Problem attempted but not solved

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Poor performance. Dice score and individual class results not provided for model evaluation</td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. UNet3D (Normal Difficulty)</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Segmentation output was not properly processed for output display</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Missing. </td><td>-3</td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-13</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally
